src/Native/System.Security.Cryptography.Native/pal_cms.cpp

```diff
@@ -103,3 +105,87 @@ extern "C" int CryptoNative_CmsGetRecipientStackFieldCount(CmsRecipientStack* re
 
     return sk_CMS_RecipientInfo_num(recipientStack);
 }
+
+extern "C" CMS_ContentInfo* CryptoNative_CmsInitializeEnvelope(ASN1_OBJECT* algorithmOid, int32_t* status)
+{
+    if (algorithmOid == nullptr || status == nullptr)
+    {
+        *status = -1;
+        return nullptr;
+    }
+
+    const EVP_CIPHER* cipher = EVP_get_cipherbyobj(algorithmOid);
+
+    if (cipher == nullptr)
+    {
+        *status = 0;
+        return nullptr;
+    }
+
+    *status = 1;
+    return CMS_EnvelopedData_create(cipher);
+}
+
+extern "C" int CryptoNative_CmsAddSkidRecipient(CMS_ContentInfo* cms, X509* cert)
+{
+    if (cms == nullptr || cert == nullptr)
+    {
+        return -1;
+    }
+
+    // TODO (3334): here we must evaluate if there's an explicit skid, or we must generate
+    // one according to the fallback behavior seen in CryptoApi. Eg: The SHA-1 hash of the publicKeyInfo
+    // blob of the encoded certificate.
+
+    return (CMS_add1_recipient_cert(cms, cert, CMS_USE_KEYID) == nullptr) ? 0 : 1;
+}
+
+extern "C" int CryptoNative_CmsAddIssuerAndSerialRecipient(CMS_ContentInfo* cms, X509* cert)
+{
+    if (cms == nullptr || cert == nullptr)
+    {
+        return -1;
+    }
+
+    return (CMS_add1_recipient_cert(cms, cert, 0) == nullptr) ? 0 : 1;
+}
+
+extern "C" int CryptoNative_CmsAddOriginatorCert(CMS_ContentInfo* cms, X509* cert)
+{
+    if (cms == nullptr || cert == nullptr)
+    {
+        return -1;
+    }
+
+    return CMS_add1_cert(cms, cert);
+}
+
+extern "C" int CryptoNative_CmsCompleteMessage(CMS_ContentInfo* cms, BIO* data, bool detached)
+{
+    if (cms == nullptr || data == nullptr)
+    {
+        return -1;
+    }
+
+    // OpenSSL supports detached content from the CMS_ContentInfo structure using the CMS_DETACHED
+    // flag, however we don't support this feature in Core. This function will allocate a new ASN1_OCTET_STRING
+    // to store the message's content and flag it as a created octet string instead of one that was read in.
+    // If this function is not used, the content won't get attached to the CMS_ContentInfo structure and the
+    // i2d function will return the encoding for a CMS with no content inside of it.
+    if (!detached && CMS_set_detached(cms, 0) == 0)
```

Comment:
The OpenSSL implementations of CMS_sign and CMS_encrypt call CMS_set_detached(cms, 0) only when the detached flag is not set, if the flag is set to true, and CMS_set_detached(cms, detached) is called this would free the content contained in the CMS_ContentInfoStructure, which is something that the functions in it don't normally do although it makes sense.